src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2805,13 +2827,24 @@ internal void CheckAndReportRequiresUnreferencedCode (MethodDefinition method)
 			if (ShouldSuppressAnalysisWarningsForRequiresUnreferencedCode ())
 				return;
 
-			if (Annotations.TryGetLinkerAttribute (method, out RequiresUnreferencedCodeAttribute requiresUnreferencedCode)) {
-				string formatString = SharedStrings.RequiresUnreferencedCodeMessage;
-				string arg1 = MessageFormat.FormatRequiresAttributeMessageArg (requiresUnreferencedCode.Message);
-				string arg2 = MessageFormat.FormatRequiresAttributeUrlArg (requiresUnreferencedCode.Url);
-				string message = string.Format (formatString, method.GetDisplayName (), arg1, arg2);
-				_context.LogWarning (message, 2026, currentOrigin, MessageSubCategory.TrimAnalysis);
+			if (method.IsStatic || method.IsConstructor) {
```

Comment:
We might need to think about how this applies to static constructors.

Is this going to warn anywhere?

```csharp
[RUC]
class Foo
{
    public static object Instance;

    static Foo()
    {
        Instance = Activator.CreateInstance(Type.GetType(WebClient.DownloadString("https//example.com/")));
    }
}

class Program
{
    static void Main() => Console.WriteLine(Foo.Instance.ToString());
}
```

The more I think about it, the more I feel we should disallow placing RUC on static constructors because it makes things _too damn complicated_.

Reply:
Given that we can now put RUC on the type itself - the value of supporting it on static .ctor seems questionable. The RUC on static .ctor should produce warnings in almost all the cases where RUC on type will (I know that there are some potential exceptions, where the runtime is allowed to call a static method without calling the static .ctor). From a UX perspective I don't see much of a difference.

I think it should be fine to remove support for RUC on static .ctor once we have this merged. Note that we should see how many cases in libraries use RUC on static .ctor today.

Reply:
I would expect the above to warn in `Main` - that said this is something which currently probably doesn't work and we'll have to add that (check for RUC on declaring type when accessing static fields). Obsolete already handles this as expected (access to any static member issues the warning).

Reply:
I agree that once we have `RUC` on types, it makes a lot less sense and is a lot more complicated to make it work for static constructors.

Reply:
Placing Obsolete on a static constructor doesn't seem to do anything. It's not a behavior we can be okay with for RUC.

I was more thinking along the lines of:

* RUC on a class doesn't apply to the cctor (it will not suppress anything within the cctor)
* Placing RUC on a cctor is always a warning

We cannot deprecate RUC on a cctor in favor of RUC on a class because we don't allow RUC on structs and interfaces but those can have cctors.

Allowing RUC on a cctor complicates things by _a lot_. I'm not sure we handle them:

```csharp
// Where do we warn?
((IFoo)new A()).Frob();

interface IFoo
{
    static IFoo
    {
        // This code will run when the default implementation on IFoo is used
    }
    void Frob() { }
}

interface IBar : IFoo
{
    static IBar()
    {
        // This will run when the default implementation is used
    }
    void IFoo.Frob() { }
}

class A : IFoo { }
```

```csharp
// Where do we warn?
_ = Foo.MyField;

// Where do we warn?
typeof(Foo).GetField("MyField").GetValue(null);

class Foo
{
    static Foo()
    {
        // Will run if a field is accessed
    }

    public static int MyField;
}
```

Of course there will be a customer who really really wants to put this on a cctor. We tell them to lazy-initialize using their own devices.

Reply:
> We cannot deprecate RUC on a cctor in favor of RUC on a class because we don't allow RUC on structs and interfaces but those can have cctors.

Why do you think it's important to support RUC on .cctor for struct/interface? I'm sure I can come up with a reason, but at the same time, I think I would be OK simply not supporting that at all.
I personally would go for the "simple" solution - RUC on classes (as per this PR), warn on RUC on .cctor (everywhere) and ignore it.

> RUC on a class doesn't apply to the cctor (it will not suppress anything within the cctor)

The only case I think of why this might be a good thing is if there's a case where the .cctor is executed by we won't be able to warn due to access to the type. But I can't think of such a case... (doesn't mean it's not there)

Reply:
Yes, I'm trying to go for the simple solution as well - the above was my proposal for that:

* Disallow explicit RUC on cctor.
* RUC on a class doesn't imply RUC behavior on it's cctor.

Reply:
> RUC on a class doesn't imply RUC behavior on it's cctor.

I assume this means "don't do anything special about .cctor". I still don't understand why we would not want to auto-suppress warnings from within the .cctor if the type is RUC annotated (we will do that for all methods on the type).

Reply:
> I assume this means "don't do anything special about .cctor". I still don't understand why we would not want to auto-suppress warnings from within the .cctor if the type is RUC annotated (we will do that for all methods on the type).

Because it's equivalent to allowing RUC on cctor itself and brings all the problems listed above (and more).

It would suppress the warning about Type.GetType here.

```csharp
[RUC]
class Foo
{
    public static object Instance;

    static Foo()
    {
        Instance = Activator.CreateInstance(Type.GetType(WebClient.DownloadString("https//example.com/")));
    }
}

class Program
{
    static void Main() => Console.WriteLine(Foo.Instance.ToString());
}
```

Here are two options to solve this:

* Don't let the class RUC apply to the static constructor. People get a warning in the static constructor and have to deal with it.
* Let the RUC apply to static constructor as well. Now ILLink is on the hook to analyze access to static fields as well. We also have to make sure warnings are generated when reflection-accessing those fields. Feels like a bunch of work.

Reply:
Would you also not suppress warnings from static field initializers?

Reply:
Even without `RUC` on static constructors, the `Obsolete` behavior is to warn on access to any static member, including a static field:

```C#
public class C {
    public void M() {
        int x = D.X; // warning
    }
}

[Obsolete]
public class D
{
    public static int X;
}
```

I don't think changing the static constructor behavior saves us from implementing support for checking static field access.

Reply:
> I don't think changing the static constructor behavior saves us from implementing support for checking static field access.

Static field access doesn't _have to_ warn. RUC is different than Obsolete in the sense that we're declaring the class is doing something trimming incompatible. But field access is normally not trimming incompatible.

It's also tricky to map C#'s warning semantics to IL - Roslyn is actually always warning about the type - not about the member. That's why it also warns on a cast to the type, or warns when declaring a local. We're not going to warn on those I assume.

Reply:
Agreed on the concept, but if the actual action that's creating the warning is accessing the type (which triggers running the static constructor), then the warning is in the appropriate place.

Reply:
> Roslyn is actually always warning about the type - not about the member. That's why it also warns on a cast to the type, or warns when declaring a local. We're not going to warn on those I assume.

What would be the problem with warning for these patterns? It doesn't seem like you can do anything too useful with just casts/local declarations, unless you also try to instantiate the type or call static methods.

Reply:
> What would be the problem with warning for these patterns? It doesn't seem like you can do anything too useful with just casts/local declarations, unless you also try to instantiate the type or call static methods.

It will produce a lot more _unnecessary_ warnings.

I think there's a lot of value in avoiding too much unnecessary noise. It we can avoid generating a warning for something that is not actually unsafe, we should. The point of the warnings it to prevent different behavior after trimming, not "to be consistent with `ObsoleteAttribute`". If the thing illink is warning about can __under no circumstances__ cause a runtime difference, that warning is useless and noise for the developer.

(That's why I'm also asking about the static fields.)

Reply:
> Don't let the class RUC apply to the static constructor. People get a warning in the static constructor and have to deal with it.

The main problem is this. How do we expect people to deal with calls to RUC methods in static initializers? It seems like it would imply a significant code change, mostly to reduce our implementation complexity.

Reply:
> The main problem is this. How do we expect people to deal with calls to RUC methods in static initializers? It seems like it would imply a significant code change, mostly to reduce our implementation complexity.

Yup, same as we scoped out RUC on structs and interfaces. I don't even know how we would be able to support the semantics of "cctor is considered RUC" there (the problem is that for structs and interfaces, the cctor is going to run when virtual methods get invoked - Equals/GetHashCode/ToString/interface methods - so we would have to mark as RUC those too; but that immediately causes a warning because the RUC annotation is inconsistent across virtuals and we can't guarantee it would warn - we can ignore the problem for classes because the cctor is triggered from the ctor there). I'm sure we would come up with some solution. But also it's very complex.

If we say RUC on struct and interface cctors is never going to work (we're also going to block it if people explicily add it to their cctor), we are already blocking a scenario because of complexity. People will have to work around if they hit it. We can document it.

Allowing RUC on class cctors (either explicityly placed on the cctor or propagated from the class-level cctor) is also pretty complex if we want to do it in a way that it doesn't produce meaningless warnings. The workarounds in the absence of that RUC handling are the same as for interfaces and structs.